### PR TITLE
Add upsert method in state for subnets domain

### PR DESCRIPTION
**Pull request by nvinuesa**
This patch adds a `UpsertSubnets` method, which takes a list of subnets to add or update (if they already exist).

This method is necessary for the upcoming space service layer, which has the method SaveProviderSubnets. This method is currently not atomic on our current mongodb state, but the idea is to turn it atomic by running all the subnet creations/updates under one single transaction. Thus, the UpsertSubnets method will be used.

Note that the `AddSubnet` and `SubnetUpdate` methods are updated so the mutual code is extracted.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Since this is state layer is not still wired-up, this should suffice:
```
go test github.com/juju/juju/domain/network/... -gocheck.v
```

## Links

**Jira card:** JUJU-5105

### CODE REVIEW

File: domain/network/state/subnet.go [hunk 1]
```diff
@@ -20,10 +20,51 @@ const (
...
+// transaction.
+func (st *State) UpsertSubnets(ctx context.Context, subnets []network.SubnetInfo) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		for _, subnet := range subnets {
+			err := updateSubnetSpaceIDTx(
```
Comment (SimonRichardson) [inline domain/network/state/subnet.go:33]: drop the Tx suffix (addSubnet) doesn't have one.

File: domain/network/state/subnet.go [hunk 2]
```diff
@@ -20,10 +20,51 @@ const (
...
+					tx,
+					string(subnet.ID),
+					subnet.CIDR,
+					subnet.ProviderId,
+					subnet.ProviderNetworkId,
+					subnet.VLANTag,
+					subnet.AvailabilityZones,
+					subnet.SpaceID,
+					subnet.FanInfo,
+				); err != nil {
```
Comment (SimonRichardson) [inline domain/network/state/subnet.go:55]:
```suggestion
				if err := addSubnet(
					ctx,
					tx,
					subnet,
				); err != nil {
```

Comment (SimonRichardson) [reply domain/network/state/subnet.go:55]: This is doing way too much.

File: domain/network/state/subnet.go [hunk 3]
```diff
@@ -80,103 +116,135 @@ VALUES (?, ?)`
...
+	ctx context.Context,
+	uuid string,
+	cidr string,
+	providerID network.Id,
+	providerNetworkID network.Id,
+	VLANTag int,
+	availabilityZones []string,
+	spaceUUID string,
+	fanInfo *network.FanCIDRs,
+) error {
```
Comment (SimonRichardson) [inline domain/network/state/subnet.go:229]: There are too many parameters here. Make an intermediate type, you can then pass that into addSubnet as well.

File: domain/network/state/subnet.go (hunk 1 above)
Comment (nvinuesa) [reply domain/network/state/subnet.go]: it was a bad replace probably, fixed

File: domain/network/state/subnet.go (hunk 2 above)
Comment (nvinuesa) [reply domain/network/state/subnet.go]: fixed

File: domain/network/state/subnet.go [hunk 4]
```diff
@@ -20,31 +20,53 @@ const (
...
+	if subnet.FanInfo != nil {
 		subnetType = subnetTypeFanOverlaySegment
 	}
 	var spaceUUIDValue any
-	spaceUUIDValue = spaceUUID
-	if spaceUUID == "" {
-		spaceUUIDValue = nil
+	spaceUUIDValue = subnet.SpaceID
+	if subnet.SpaceID == "" {
+		spaceUUIDValue = network.AlphaSpaceId
```
Comment (SimonRichardson) [inline domain/network/state/subnet.go:69]: You can remove the following line `var spaceUUIDValue any` and just do the following:

```suggestion
	spaceUUIDValue := subnet.SpaceID
	if subnet.SpaceID == "" {
		spaceUUIDValue = network.AlphaSpaceId
```

File: domain/network/state/subnet.go [hunk 5]
```diff
@@ -80,103 +102,121 @@ VALUES (?, ?)`
...
+	); err != nil {
+		return errors.Trace(err)
+	}
+
+	if subnetType == subnetTypeFanOverlaySegment {
+		// Retrieve the underlay subnet uuid.
+		var underlaySubnetUUID string
+		row := tx.QueryRowContext(ctx, retrieveUnderlaySubnetUUIDStmt, subnet.FanInfo.FanLocalUnderlay)
+		if err := row.Scan(&underlaySubnetUUID); err != nil {
 			return errors.Trace(err)
```
Comment (SimonRichardson) [inline domain/network/state/subnet.go:124]: Can we annotate the errors in this function, it's going to be hard to know which one fails easily.

File: domain/network/state/subnet.go [hunk 6]
```diff
@@ -80,103 +102,121 @@ VALUES (?, ?)`
...
 		}
-		for _, az := range availabilityZones {
-			// Retrieve the availability zone.
-			var azUUIDStr string
-			row := tx.QueryRowContext(ctx, retrieveAvailabilityZoneStmt, az)
-			err := row.Scan(&azUUIDStr)
-			if err != nil && err != sql.ErrNoRows {
+		// If it doesn't exist, add the availability zone.
+		if err == sql.ErrNoRows {
```
Comment (SimonRichardson) [inline domain/network/state/subnet.go:175]:
```suggestion
		if errors.Is(err, sql.ErrNoRows) {
```

File: domain/network/state/subnet.go [hunk 7]
```diff
@@ -80,103 +102,121 @@ VALUES (?, ?)`
...
+	// subnets mapping table.
+	if _, err := tx.ExecContext(
+		ctx,
+		insertSubnetProviderNetworkSubnetStmt,
+		pnUUID.String(),
+		subnet.ID,
+	); err != nil {
+		return errors.Trace(err)
+	}
+	for _, az := range subnet.AvailabilityZones {
```
Comment (SimonRichardson) [inline domain/network/state/subnet.go:166]: Can we break this for loop off from here into a separate method... the addSubnet function is getting rather long.

File: domain/network/state/subnet.go (hunk 5 above)
Comment (nvinuesa) [reply domain/network/state/subnet.go]: sure, I thought I had already done that

File: domain/network/state/subnet.go (hunk 7 above)
Comment (nvinuesa) [reply domain/network/state/subnet.go:174]: sure, I made a new method only for adding availability zones, should be prettier now

**Review by SimonRichardson — COMMENTED** [review]
Last thing and then I'm happy.

File: domain/network/state/subnet.go [hunk 8]
```diff
@@ -70,6 +91,73 @@ VALUES (?, ?)`
...
+	); err != nil {
+		return errors.Annotatef(err, "inserting association between provider network id %q and subnet %q", subnet.ProviderNetworkId, subnet.ID)
+	}
+
+	return addAZs(ctx, tx, subnet)
+}
+
+// addAZs adds the availability zones of a subnet if they don't exist, and
+// update the availability_zone_subnet table with the subnet's id.
+func addAZs(ctx context.Context, tx *sql.Tx, subnet network.SubnetInfo) error {
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply domain/network/state/subnet.go]: done

**Review by SimonRichardson — APPROVED** [review]

**Pull request merged**